### PR TITLE
fix(frontend): order same-timestamp send/receive pairs (receive above send)

### PR DESCRIPTION
**Pull request by sbpublic**
# Motivation

In the activity list, the two legs of a single operation — a swap's send + receive, or a self-transfer — share the same block timestamp. `sortTransactions` compared only the (second-normalized) timestamp and returned `0` on ties, so these pairs fell back to `Array.sort` insertion order (driven by token/provider load order). The result looked arbitrary: sometimes the send appeared above the receive, sometimes below.

# Changes

- `sortTransactions`: when timestamps are equal, break the tie deterministically by type — the received leg sorts above the sent one. This is the single comparator the whole activity list uses, so every same-second send/receive pair (swaps, self-transfers, …) now orders consistently.
- `sortTransactions`: return `0` when both timestamps are nullish (previously `1` — and `1` again when swapped, an antisymmetry-contract violation that left such pairs in engine-dependent order). The single-nullish case still sorts nullish last. (From Copilot review.)

# Tests

- Added `sortTransactions` tests: the received leg sorts above the sent one for a same-timestamp pair (regardless of input order), and the comparator returns `0` for two nullish-timestamp transactions in both argument orders.
- `transactions.utils.spec.ts` passes in full (78).
- Manual verification on `test_fe_3`.

> Independent of #13242 — based directly on `main` and mergeable on its own. This is a standalone determinism fix for same-timestamp send/receive ordering (it also covers self-transfers). The full _visible_ swap-pair effect additionally needs #13242's dedup fix (which keeps both legs); the two can land in any order.

---

Model: Claude Opus 4.8 (`claude-opus-4-8`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/frontend/src/lib/utils/transactions.utils.ts [hunk 1]
```diff
@@ -319,18 +319,38 @@ const isMicroTransaction = ({
...
+
+		// The two legs of one operation (a swap, or a self-transfer) share the same block timestamp
+		// and tie here. Break the tie deterministically — received leg above the sent one — so these
+		// pairs render consistently instead of in an arbitrary insertion order.
+		return bySeconds !== 0
+			? bySeconds
+			: sameTimestampTypeRank(typeA) - sameTimestampTypeRank(typeB);
 	}
 
 	return nonNullish(timestampA) ? -1 : 1;
```
[the remark this answers is not in the document]
Comment (sbpublic) [reply src/frontend/src/lib/utils/transactions.utils.ts]: Good catch — fixing the comparator contract for the nullish/nullish case.

**Summary**: when both timestamps are nullish, `return nonNullish(timestampA) ? -1 : 1` returns `1`, and it returns `1` again with the arguments swapped — so `compare(a, b)` and `compare(b, a)` are both positive.

**Analysis**: that breaks the comparator's antisymmetry contract, leaving the relative order of two nullish-timestamp transactions engine-dependent — which undercuts the determinism this PR adds. The single-nullish cases are already correct (nullish sorts last).

**Severity**: low (both-nullish is rare, but it's a real contract violation with non-deterministic ordering).

**Fix**: return `0` when both timestamps are nullish, keeping nullish-last for the single-nullish case — `if (nonNullish(timestampA)) { return -1; } return nonNullish(timestampB) ? 1 : 0;`.

🤖 Claude Opus 4.8

File: src/frontend/src/tests/lib/utils/transactions.utils.spec.ts [hunk 2]
```diff
@@ -810,6 +810,19 @@ describe('transactions.utils', () => {
...
+			const received = { timestamp: 5, type: 'receive' } as AnyTransactionUi;
+
+			// Deterministic regardless of input order: the received leg always ends up above the sent one.
+			expect(
+				[sent, received].sort((a, b) => sortTransactions({ transactionA: a, transactionB: b }))
+			).toEqual([received, sent]);
+			expect(
+				[received, sent].sort((a, b) => sortTransactions({ transactionA: a, transactionB: b }))
+			).toEqual([received, sent]);
+		});
```
[the remark this answers is not in the document]
Comment (sbpublic) [reply src/frontend/src/tests/lib/utils/transactions.utils.spec.ts:799]: Good catch — adding the nullish/nullish regression.

**Summary**: no test covered `sortTransactions` when both timestamps are nullish.

**Analysis**: that's exactly the case the comparator-contract fix (the related line-356 comment) addresses, so without a test a regression to the old `1`/`1` return would slip through uncaught. The existing nullish test only exercises the single-nullish, sorts-last path.

**Severity**: low (edge-case coverage gap).

**Fix**: add the suggested test asserting `sortTransactions` returns `0` for two nullish-timestamp transactions in both argument orders — adopting Copilot's changeset verbatim.

🤖 Claude Opus 4.8

**Review by DenysKarmazynDFINITY — APPROVED** [review]
